### ISSUE TITLE
Rule Proposal: no-array-destructured-return

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#6054]
From [disallowArrayDestructuringReturn](http://jscs.info/rule/disallowArrayDestructuringReturn).

This rule will prevent to receive returned values with array destructuring as the favor of named returned values.
I think this rule should allow array destructuring with rest elements: `let [head, ...rest] = foo();`

``` js
{
    "no-nameless-multiple-return": "error"
}
```

**Valid:**

``` js
let {left, right} = foo();
let [head, ...rest] = foo();
```

**Invalid:**

``` js
let [a, b] = foo();
```

**Comment by nzakas** [comment]
I'm not sure the name makes much sense. I'd see something like `no-array-destructured-return` as being easier to understand.

**Comment by kaicataldo** [comment]
@eslint/eslint-team Anyone want to champion this? I would but I have two other rules on my plate at the moment and don't want to commit to more than I can handle. 

@mysticatea I agree with @nzakas on the name. Curious why you feel like array elements with rest elements should be allowed, since they weren't in the original JSCS rule.

**Comment by mysticatea** [comment]
I thought the purpose of the JSCS rule. There is a hint in the code example. I thought the purpose of this rule is to disallow an array as multi-value returning because:

- Object destructuring would clarify the meaning of returned values more than array destructuring.
- Arrays should be used for items of an unspecified number.

So, this rule disallows array destructuring of **fixed size**, which is receiving returned values.

Rest elements are not fixed size.

**Comment by mikesherov** [comment]
I agree with @mysticatea here. The rest example written in ES5 is:

```
var rest = foo();
var head = rest.shift();
```

Which is to say, it's a perfectly valid array assignment then manipulation in a single step, where the elements are presumably of the same type. The invalid example in es5 is:

```
var arr = foo();
var a = arr[0];
var b = arr[1];
```

Which is using an array as a poor man's object to store values of presumably different type. 

This lint exists essentially to follow https://sourcemaking.com/refactoring/replace-array-with-object

**Comment by vitorbal** [comment]
I'm willing to champion and implement this rule. I also agree with @mysticatea that rest elements should be allowed. @kaicataldo considering @mysticatea's and @mikesherov thoughts above, do you feel strongly against allowing rest elements?

**Comment by kaicataldo** [comment]
@vitorbal Nope! Was more curious than anything.

**Comment by vitorbal** [comment]
I started implementing this. I should have a PR ready some time next week!

**Comment by ljharb** [comment]
I agree on changing the name. JS does not have multiple return - it only ever has single return. Destructuring is a caller-side shortcut that unpacks a single array or object.

**Comment by kaicataldo** [comment]
@jlharb Agreed - changing the name of the issue to reflect @nzakas's proposed name, since it looks like we have support for that

**Comment by kaicataldo** [comment]
The TSC decided in the 2017-04-13 meeting not to include this rule in core. @not-an-aardvark will open a separate issue to discuss an alternative approach that would check return statements instead.
